source/agora/common/MerkleTree.d

```diff
@@ -0,0 +1,345 @@
+
+    public inout(MerkleNode)* searchNode (Hash hash) inout
+    {
+        if (this.hash == hash) return &this;
+
+        if (auto res = this.left.searchNode(hash))
+            return res;
+
+        return this.right.searchNode(hash);
+    }
+
+    /*******************************************************************************
+
+        Get Merkle path
+
+        Params:
+            hash = `Hash` of `Transaction`
+
+        Returns:
+            Return array of `Hash` of Merkle path
+
+    *******************************************************************************/
+
+    public Hash[] getMerklePath (Hash hash)
+    {
+        bool getPath (MerkleNode* node, Hash hash, ref Hash [] path)
+        {
+            if (node is null) return false;
+
+            if (node.hash == hash) return true;
+
+            if (getPath(node.right, hash, path))
+            {
+                path ~= node.left.hash;
+                return true;
+            }
+
+            if (getPath(node.left, hash, path))
+            {
+                path ~= node.right.hash;
+                return true;
+            }
+
+            return false;
+        }
+
+        Hash [] path;
+        getPath(&this, hash, path);
+        return path;
+    }
+
+    /*******************************************************************************
+
+        Prove merkle path
+
+        Params:
+            hash = `Hash` of `Transaction`
+            path = `Hash` of Merkle path
+            root = `Hash` of Merkle root
+            index = Index of the hash in the array of transactions. It starts at zero.
+
+        Returns:
+            Return true if the Merkle proof is completed. Return false otherwise.
+
+    *******************************************************************************/
+
+    public static bool proveMerklePath (Hash hash, const Hash[] path, Hash root, size_t index)
+    {
+        if (path.length == 0) return false;
+
+        foreach (ref p; path)
+        {
+            if (index % 2)
+                hash = mergeHash(p, hash);
+            else
+                hash = mergeHash(hash, p);
+
+            index >>= 1;
+        }
+
+        return (hash == root);
+    }
+
+    /***************************************************************************
+
+        Implements hashing support
+
+        Params:
+            dg = hashing function
+
+    ***************************************************************************/
+
+    public void computeHash (scope HashDg dg) const nothrow @safe @nogc
+    {
+        // Our node contains either a transaction, in which case it is a leaf node,
+        // or 2 childs (which might be the same if the tree is unbalanced)
+        if (this.tx !is null)
+            hashPart(*this.tx, dg);
+        else
+        {
+            hashPart(this.left.hash, dg);
+            hashPart(this.right.hash, dg);
+        }
+    }
+}
+
+/// Merge two hash
+public Hash mergeHash (Hash h1, Hash h2)
```

Comment:
Will this function be used outside this module? If not, it should be `private`. Also it's missing parameter documentation.